src/czar/CzarFamilyMap.cc

```
@@ -0,0 +1,263 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "czar/CzarFamilyMap.h"
+
+// System headers
+#include <sstream>
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+// Qserv headers
+#include "qmeta/QMeta.h"
+#include "cconfig/CzarConfig.h"
+#include "czar/Czar.h"
+#include "czar/CzarRegistry.h"
+#include "qmeta/Exceptions.h"
+#include "util/Bug.h"
+#include "util/TimeUtils.h"
+
+using namespace std;
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.czar.CzarFamilyMap");
+}  // namespace
+
+namespace lsst::qserv::czar {
+
+CzarFamilyMap::Ptr CzarFamilyMap::create(std::shared_ptr<qmeta::QMeta> const& qmeta) {
+    // There's nothing the czar can do until with user queries until there's been at least
+    // one successful read of the database family tables, as the czar doesn't know where to find anything.
+    Ptr newPtr = nullptr;
+    while (newPtr == nullptr) {
+        try {
+            newPtr = Ptr(new CzarFamilyMap(qmeta));
+        } catch (ChunkMapException const& exc) {
+            LOGS(_log, LOG_LVL_WARN, "Could not create CzarFamilyMap, sleep and retry " << exc.what());
+        }
+        if (newPtr == nullptr) {
+            this_thread::sleep_for(10s);
+        }
+    }
+
+    return newPtr;
+}
+
+CzarFamilyMap::CzarFamilyMap(std::shared_ptr<qmeta::QMeta> const& qmeta) : _qmeta(qmeta) {
+    try {
+        auto mapsSet = _read();
+        if (!mapsSet) {
```

The result of the function doesn't have to be stored in a separate variable, as it's only used to check if the read failed:
```
if (!_read()) {
```

My second comment on the method `_read()` is that it should throw an appropriate exception to indicate the abnormal situation instead of returning the status code.